src/db/spanner/batch.rs

```diff
@@ -269,22 +253,74 @@ pub async fn do_append_async(
     //  ("<fxa_uid>", "<fxa_kid>", 101, "ba1", "bso2", NULL, "payload2", NULL)]
     // https://cloud.google.com/spanner/docs/structs#creating_struct_objects
     let mut running_size: usize = 0;
-    let rows: Vec<_> = bsos
-        .into_iter()
-        .map(|bso| {
-            let sortindex = bso
-                .sortindex
-                .map(|sortindex| as_value(sortindex.to_string()))
-                .unwrap_or_else(null_value);
-            let payload = bso.payload.map(as_value).unwrap_or_else(null_value);
-            if payload != null_value() {
-                running_size += payload.get_string_value().len();
-            }
-            let ttl = bso
-                .ttl
-                .map(|ttl| as_value(ttl.to_string()))
-                .unwrap_or_else(null_value);
 
+    // problem: Append may try to insert a duplicate record into the batch_bsos table.
+    // this is because spanner doesn't do upserts. at all. Batch_bso is a temp table and
+    // items are eventually rolled into bsos.
+
+    fn exist_idx(collection_id: &str, batch_id: &str, bso_id: &str) -> String {
+        format!(
+            "{collection_id}::{batch_id}::{bso_id}",
+            collection_id = collection_id,
+            batch_id = batch_id,
+            bso_id = bso_id,
+        )
+    }
+
+    struct UpdateRecord {
+        sortindex: Value,
+        ttl: Value,
+        bso_id: String,
+        payload: String,
+    };
+
+    //prefetch the existing batch_bsos for this user.
+    let mut existing = HashSet::new();
+    let mut existing_stream = db
+    .sql("SELECT collection_id, batch_id, batch_bso_id from batch_bsos where fxa_uid=@fxa_uid and fxa_kid=@fxa_kid;")?
```

Comment:
I think specifying the bso_ids on hand here (like `post_bsos` does, i.e. `AND batch_bso_id IN UNNEST(@ids`) is worth the effort.

We'd specify a max number of 100 bso_ids (on hand) vs selecting all might potentially return 1666 bso_ids. That doesn't sound like the greatest payoff but isolating the SELECT to fewer rows might be better for the Spanner r/w transaction (the less rows/columns it has to lock the better)